torch/csrc/distributed/c10d/comm.hpp

```diff
@@ -81,38 +81,38 @@ class TORCH_API GradBucket {
 // Base class of both `PythonCommHook` and `CppCommHook`.
 // Requires implementing 1) `runHook` method that communicates gradients
 // asynchronously, and 2) `parseHookResult` method that converts the hook
-// result into a tensor vector.
+// result into a tensor.
 class TORCH_PYTHON_API CommHookInterface {
  public:
-  virtual ~CommHookInterface() {}
+  virtual ~CommHookInterface() = default;
 
   // Passes the input grad bucket to the registered communication hook.
-  // Once the tensors in the bucket are ready, kicks off the hook asynchronously
+  // Once the tensor in the bucket are ready, kicks off the hook asynchronously
   // and returns a future that holds the communication results.
   virtual c10::intrusive_ptr<c10::ivalue::Future> runHook(
       GradBucket& bucket) = 0;
 
-  // Returns the resulting tensors once the communication hook result is
-  // ready. The resulting tensors will then be copied to the grads of
+  // Returns the resulting tensor once the communication hook result is
+  // ready. The resulting tensor will then be copied to the grads of
   // individual parameters.
-  virtual std::vector<at::Tensor> parseHookResult(
+  virtual at::Tensor parseHookResult(
       const c10::IValue& result) = 0;
 };
 
 namespace detail {
 // This helper function is called both by CppCommHookInterface below and inside
 // reducer.
-inline std::vector<at::Tensor> parseCppCommHookResult(
+inline at::Tensor parseCppCommHookResult(
     const c10::IValue& result) {
   TORCH_INTERNAL_ASSERT(
       result.isTensor() || result.isTensorList(),
       "expected the hook result is either a Tensor or a TensorList");
 
   if (result.isTensor()) {
-    return {result.toTensor()};
+    return result.toTensor();
   }
 
-  return result.toTensorVector();
+  return result.toTensorVector()[0];
```

Comment:
Can cpp comm. hook somehow still return a tensor list? If not, should we be able to remove this block entirely and expect `result` to always be a tensor?

Reply:
Good observation! This can be refactored in the next PR. Currently only `allreduce_hook` which directly calls `pg.allreduce` still returns a tensor list. I need to confirm with @lw if it's OK to do `dist.all_reduce(tensor).get_future().wait()[0]`.